compiler/rustc_typeck/src/check/upvar.rs

```diff
@@ -527,6 +534,182 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+            }
+            reasons = format!("{}drop order", reasons);
+        }
+
+        reasons
+    }
+
+    /// Figures out the list of root variables (and their types) that aren't completely
+    /// captured by the closure when `capture_disjoint_fields` is enabled and auto-traits
+    /// differ between the root variable and the captured paths.
+    ///
+    /// The output list would include a root variable if:
+    /// - It would have been captured into the closure when `capture_disjoint_fields` wasn't
+    ///   enabled, **and**
+    /// - It wasn't completely captured by the closure, **and**
+    /// - One of the paths captured does not implement all the auto-traits its root variable
+    ///   implements.
+    fn compute_2229_migrations_for_trait(
+        &self,
+        min_captures: Option<&ty::RootVariableMinCaptureList<'tcx>>,
+        _closure_clause: hir::CaptureBy,
+        var_hir_id: hir::HirId,
+    ) -> Option<FxHashSet<&str>> {
+        use crate::rustc_middle::ty::ToPredicate;
+        use crate::rustc_middle::ty::WithConstness;
+        use rustc_trait_selection::traits::query::evaluate_obligation::InferCtxtExt;
+
+        let root_var_min_capture_list = if let Some(root_var_min_capture_list) =
+            min_captures.and_then(|m| m.get(&var_hir_id))
+        {
+            root_var_min_capture_list
+        } else {
+            return None;
+        };
+
+        let ty = self.infcx.resolve_vars_if_possible(self.node_ty(var_hir_id));
+
+        let tcx = self.infcx.tcx;
+
+        let cause = ObligationCause::misc(self.tcx.hir().span(var_hir_id), self.body_id);
+
+        // <ty as Clone>
+        let clone_trait_ref = TraitRef {
+            def_id: tcx.lang_items().clone_trait().unwrap(),
+            substs: tcx.mk_substs_trait(ty, &[]),
+        };
+
+        let clone_obligation_root = Obligation::new(
+            cause.clone(),
+            self.param_env,
+            clone_trait_ref.without_const().to_predicate(tcx),
+        );
+
+        // <ty as Sync>
+        let sync_trait_ref = TraitRef {
+            def_id: tcx.lang_items().sync_trait().unwrap(),
+            substs: tcx.mk_substs_trait(ty, &[]),
+        };
+
+        let sync_obligation_root = Obligation::new(
+            cause.clone(),
+            self.param_env,
+            sync_trait_ref.without_const().to_predicate(tcx),
+        );
+
+        // <ty as Unpin>
+        let unpin_trait_ref = TraitRef {
+            def_id: tcx.lang_items().unpin_trait().unwrap(),
+            substs: tcx.mk_substs_trait(ty, &[]),
+        };
+
+        let unpin_obligation_root = Obligation::new(
+            cause.clone(),
+            self.param_env,
+            unpin_trait_ref.without_const().to_predicate(tcx),
+        );
+
+        // Check whether catpured fields also implement the trait
+        let mut temp_reasons = FxHashSet::default();
+
+        for capture in root_var_min_capture_list.iter() {
```

Comment:
I think this function can use some form of helper, because I see a quite a fair bit of repetition.

Something like
```
// return true if the variables needs to be migrated
fn 2229_trait_migration_helper(&self, var_hir_id, trait_def_id) -> bool {
 
      // checks that if the trait is satisifed by root varirable; if not return false
     // check if the trait is not satisfied by an of the captured paths, if so return true. 
}
```
One thing to check with niko would be if we want to report why a paricular root variable is being migrated, i.e. store (variable, reason) pairs.